controller/volume_controller.go

```diff
@@ -902,7 +902,12 @@ func (vc *VolumeController) ReconcileVolumeState(v *longhorn.Volume, es map[stri
 		if err != nil {
 			return err
 		}
-		// make sure the volume is detached before automatically salvage
+		if autoSalvage && !v.Status.IsStandby && !v.Status.RestoreRequired {
```

Comment:
I don't understand why moving this code will solve the problem.

1. What triggered the problem before?
2. Why we no longer need to check if the data exists or not.
3. Why previously the logic only applies to if the node is not down or deleted?

Reply:
> What triggered the problem before?
This was reason that triggered the problem https://github.com/longhorn/longhorn/issues/1942#issuecomment-722698783. In short, Longhorn sometime succeeds to salvage replicas and save replicas CRs to `etcd` but fail to save engine to `etcd`.

> Why we no longer need to check if the data exists or not.
> Why previously the logic only applies to if the node is not down or deleted?

We still do all of this check before salvage the failed replicas in Volume Controller. 

However, I don't think that we need to do those check for before setting ` e.Spec.SalvageRequested=true` because when all replicas fail and `autoSalvage==true`, we must set ` e.Spec.SalvageRequested=true` so that engine picks the best replica the next time it starts no matter whether the current node is down or deleted, or data exists.  If we don't set `e.Spec.SalvageRequested=true`, Longhorn will pick all non-failed replicas the next time engine starts.

Reply:
Talked with Phan, `v.Status.State == types.VolumeStateDetached` works as a barrier to make sure the volume status was updated. And since engine update will be done before the volume update, so it also guarantees the engine state has been updated in the etcd.